scenarios/common/infrastructure_deployment.py

```diff
@@ -0,0 +1,131 @@
+    filename = os.path.join(os.path.dirname(__file__), deployment_hash)
+    with open(filename, 'wt') as f:
+        if success:
+            f.write(execution_id)
+        else:
+            f.write('Failed')
+
+
+def tests_finished(deployment_hash, id):
+    filename = os.path.join(os.path.dirname(__file__), deployment_hash + '.' + id)
+    with open(filename, 'wt') as f:
+        f.write('Done')
+
+
+def wait_for_infrastructure(deployment_hash):
+    filename = os.path.join(os.path.dirname(__file__), deployment_hash)
+    while True:
+        if os.path.exists(filename):
+            with open(filename, 'rt') as f:
+                global execution_id
+                execution_id = f.read()
+                if execution_id == 'Failed':
+                    return False
+                return True
+        else:
+            time.sleep(5)
+
+
+def wait_for_tests_finished(config, deployment_hash):
+    done_files = []
+    filename = os.path.join(os.path.dirname(__file__), deployment_hash)
+
+    if hasattr(config, 'workerinput'):
+        done_files = [filename + '.gw' + str(i)
+                      for i in range(1, config.workerinput['workercount'])]
+
+    while True:
+        if all(os.path.exists(file) for file in done_files):
+            break
+        else:
+            time.sleep(5)
+
+    os.remove(filename)
+    for file in done_files:
+        os.remove(file)
+
+
+def terraform_apply(working_dir, deployment_hash):
+    terraform = Terraform(working_dir=working_dir)
+    return_code, stdout, stderr = terraform.init()
+    if return_code != 0:
+        print(stdout)
+        print(stderr)
+        infrastructure_deployed(deployment_hash, False)
+        assert False
+
+    return_code, stdout, stderr = terraform.apply(skip_plan=True)
+    if return_code != 0:
+        print(stdout)
+        print(stderr)
+        infrastructure_deployed(deployment_hash, False)
+        assert False
+
+    infrastructure_deployed(deployment_hash)
+
+
+def terraform_destroy(working_dir):
+    terraform = Terraform(working_dir=working_dir)
+    return_code, stdout, stderr = terraform.destroy(force=True)
+    if return_code != 0:
+        print(stdout)
+        print(stderr)
+        assert False
+
+    # Terraform on Windows creates some hardlinks, so they should be removed first
+    # for root, dirs, files in os.walk(tmpdir, topdown=False):
+    #     for name in dirs:
+    #         if os.stat(os.path.join(root, name)).st_size == 0:
+    #             os.unlink(os.path.join(root, name))
+
+    shutil.rmtree(os.path.join(working_dir, '.terraform', 'plugins'))
+
+
+def build_master_template(request):
+    common_seen = set()
+    master_template = ''
+    deployment_info = ''
+
+    for item in request.node.items:
+        cls = item.parent._obj
+
+        if cls.common_template not in common_seen:
+            master_template += cls.common_template
+            common_seen.add(cls.common_template)

```

Comment:
common_seen doesn't seem to be used at all.

Reply:
This part will be different.. I'll just add all providers to the template every time